dbms/src/Functions/IFunction.h

```diff
@@ -277,6 +282,11 @@ class FunctionBuilderImpl : public IFunctionBuilder
         return buildImpl(arguments, getReturnType(arguments));
     }
 
+    bool isDeterministic() const override { return isDeterministicInScopeOfQuery(); }
```

Comment:
This looks somewhat wrong -- this is a generic implementation, and in general, it doesn't follow that if the function is deterministic, it is deterministic in scope of query. OTOH, the reverse is always true: if it is always deterministic, it is deterministic for the duration of query as well. Maybe we should make this an enum? As a reference, postgres has similar categories for determinism, though I'm not sure we should adopt their names: https://www.postgresql.org/docs/current/xfunc-volatility.html

Reply:
I definitely agree that using enum for function categorization would do much better here and possibly would make client code more readable. It would also let to incorporate other categories more easily (we might consider additional categories of volatility: dictionary and replica)